test/cpp/end2end/xds/xds_end2end_test.cc

```diff
@@ -1318,29 +1318,21 @@ class XdsEnd2endTest : public ::testing::TestWithParam<TestType> {
   }
 
   bool WaitForNack(
-      std::function<AdsServiceImpl::ResponseState::State()> get_state,
-      StatusCode expected_status = StatusCode::UNAVAILABLE) {
```

Comment:
It seems useful to check that the client RPCs fail with the right code in this function.  If there are certain cases where the RPCs may not fail, then how about changing the code here to call `CheckRpcSendOk()` instead of `CheckRpcSendFailure()` if `expected_status` is `OK`?

Reply:
I think it's better to separate checking the RPC status from `WaitForNack` for exactly the case that I described in the issue where the test does not know what the status is going to be.

Tests that want to make sure that RPCs fail with a certain status should explicitly check for it.

Reply:
Yes, tests should explicitly check that.  And allowing them to specify what RPC response code to expect does that.  I don't see what benefit we gain by requiring them to send a separate RPC to check this when we can just check the status code of the RPCs that we're already sending.

Moreover, even if we agreed that that was the right thing to do, removing this parameter means that all existing tests that are calling `WaitForNack()`, either directly or indirectly, are no longer verifying what they were before, so this is basically breaking existing tests.  To avoid that, you would need to look at all tests that are currently calling `WaitForNack()` and change them to verify the RPC status code separately.

Stepping back, I think the real problem here is that we're starting the server before we set the new resources on the xDS server, which means that the server may see the default resources when it starts instead of the ones that we actually want to start the test with.  If you look at the `UseOriginalDstNotSupported` test, the test itself *should* cause the RPCs to be failed with UNAVAILABLE.  The fact that that's not happening is an artifact of the strange way that we set up the test framework, so I think that's the actual problem here.

I think the right way to solve this problem is to address this TODO:

https://github.com/grpc/grpc/blob/826b888e75038d00d7d13a5679613a7de66dcb95/test/cpp/end2end/xds/xds_end2end_test.cc#L671

Each test should explicitly start up the set of backends it wants to run, which means that each test should be able to avoid starting the backend until after it has set the appropriate resources on the xDS server.